runtime/lua/vim/_editor.lua

```diff
@@ -1036,56 +1036,97 @@ end
-    major, minor = tonumber(major), tonumber(minor)
-
-    local hard_deprecated_since = string.format('nvim-%d.%d', major, minor - 1)
-    -- Assume there will be no next minor version before bumping up the major version
-    local is_hard_deprecated = minor == 0 or vim.fn.has(hard_deprecated_since) == 1
-    if not is_hard_deprecated then
-      return
-    end
-
-    local removal_version = string.format('nvim-%d.%d', major, minor)
-    will_be_removed = vim.fn.has(removal_version) == 1 and 'was removed' or will_be_removed
-  else
-    vim.validate {
-      name = { name, 'string' },
-      alternative = { alternative, 'string', true },
-      version = { version, 'string', true },
-      plugin = { plugin, 'string', true },
-    }
-  end
+  vim.validate {
+    name = { name, 'string' },
+    alternative = { alternative, 'string', true },
+    version = { version, 'string', true },
+    plugin = { plugin, 'string', true },
+  }
 
   local msg = ('%s is deprecated'):format(name)
   msg = alternative and ('%s, use %s instead.'):format(msg, alternative) or (msg .. '.')
-  msg = ('%s%s\nFeature %s in %s %s'):format(
-    msg,
-    (plugin == 'Nvim' and ' :help deprecated' or ''),
-    will_be_removed,
-    plugin,
-    version
-  )
+  msg = ('%s\nFeature will be removed in %s %s'):format(msg, plugin, version)
   local displayed = vim.notify_once(msg, vim.log.levels.WARN)
   if displayed and backtrace ~= false then
     vim.notify(debug.traceback('', 2):sub(2), vim.log.levels.WARN)
   end
   return displayed and msg or nil
 end
 
+do
+  local function truncated_echo(msg)
+    -- Truncate message to avoid hit-enter-prompt
+    local max_width = vim.o.columns * math.max(vim.o.cmdheight - 1, 0) + vim.v.echospace
+    local msg_truncated = string.sub(msg, 1, max_width)
+    vim.api.nvim_echo({ { msg_truncated, 'WarningMsg' } }, true, {})
+  end
+
+  local notified = false
+
+  function vim._truncated_echo_once(msg)
+    if not notified then
+      truncated_echo(msg)
+      notified = true
+      return true
+    end
+    return false
+  end
+end
+
+local function traceback()
+  local level = 4
+  local backtrace = { 'stack traceback:' }
+  while true do
+    local info = debug.getinfo(level, 'Sl')
+    if not info then
+      break
+    end
+    local msg = ('  %s:%s'):format(info.source:sub(2), info.currentline)
+    table.insert(backtrace, msg)
+    level = level + 1
+  end
+  return table.concat(backtrace, '\n')
+end
+
+--- Shows a deprecation message to the user. Internal function customized to fit neovim's needs.
+---
+---@param name        string     Deprecated feature (function, API, etc.).
+---@param version     string     Version when the deprecated function will be removed.
+---@param opts        table|nil  Optional parameters.
+---                              - alternative: suggested replacement function
+---
+---@return string|nil # Deprecated message, or nil if no message was shown.
+function vim._deprecate(name, version, opts)
```

Comment:
can we avoid the `_deprecate` variant, and have just one function

Reply:
How else are we going to deprecate `vim.deprecate`?

Reply:
On a more serious note, this will allow us to get the interface right _first_ before replacing the old `vim.deprecate`.

Reply:
In that case we should support the old parameters/arity, so that plugins don't need to "migrate" once we swap over the new version. We can support kwargs as an alternative, for new callers, if needed.

Reply:
We can but there'd be no point. It would essentially be two functions baked into one with a phat if/else check. The traceback argument would always be ignored for Nvim. It's just a mess in general. If you absolutely insist then I don't mind, but I really don't see the benefit.

And as clason said the current function signature is bad and not easily extensible as we don't have an opts table. I think we need to cut our losses with the current deprecation function and work on something new.

Reply:
> In that case we should support the old parameters/arity, so that plugins don't need to "migrate" once we swap over the new version. We can support kwargs as an alternative, for new callers, if needed.

Not a fan. Variable number of arguments is not a good design becausenit is not extensible.

Reply:
> I think we need to cut our losses with the current deprecation function and work on something new.

how are we going to cut our losses if we have to support both?

Reply:
I meant until we remove `vim.deprecate`.

Reply:
OK, enough lazy backseat reviewing, I need actual guidance: do we add plugin deprecations into this mechanism or not? That completely changes the solution. It would technically be breaking though if we do include plugins into this behavior.

Reply:
Fuck it, I'm reverting it then.

Reply:
I'll revert this change then.

Reply:
I've reverted it to use old function signature and to only use a single function. Are you open to "fixing" the function signature in the future to comply with our guidelines (meaning, mandatory arguments first and optional arguments in an `opts` table)? I'd hate to solidify the current function signature in all eternity, it's just not good unfortunately.